### DISCUSSION TITLE
Strategy to dynamically insert and search data

### DISCUSSION THREAD

**Discussion by ahndroo** [milvus-io/milvus#21538] [Q&A and General discussion] [answered]
Hi all! 

I have a current pipeline that is continually inserting data into a Collection.  Periodically, a 'sync' step is performed where I search a sample of the inserted data to see if any historical data are matches.  These are then grouped into a common ID on my end so I can track each identities.  For example: 
new_inserted_data_id = 1
matching_historical_data = 6,7,8
id["Some_ID"] = [1,6,7,8]

Now the problem is that to perform the dynamic search on the fly, I need to load the Collection for each sync period and this is a time consuming process.  The period update disrupts that continual data insertion and gives unacceptable latency.   

Are there any ideas for strategies where I can perform this dynamic insert and search step faster?

**Comment by yhmo** [comment, marked as answer]
No need to call collection.load() for each search.
Typical workflow for dynamic data:
- create collection
- create index
- collection.load()
- insert()
- search()
- insert()
- search()
- ......

New inserted data visibility is controlled by the parameter "consistency_level" of search(): https://milvus.io/docs/search.md
`Milvus supports setting consistency level specifically for a search. The example in this topic sets the consistency level as Strong. You can also set the consistency level as Bounded, Session or Eventually. See [Consistency](https://milvus.io/docs/consistency.md) for more information about the four consistency levels in Milvus.`

You can use "Bounded" consistency level to search:
```
results = collection.search(
	data=[[0.1, 0.2]], 
	anns_field="book_intro", 
	param=search_params,
	limit=10, 
	expr=None,
	consistency_level="Bounded"
)
```

**Reply by barzan-hayati** [reply]
I have a collection containing `500,000` rows, each of size `512`. After adding a new row to the collection, I monitor the number of entities in the collection every minute. It takes a while for the collection size to update, even though I am not performing any search operations. I also recognize that using `flush()` isn't a viable solution to speed up this process.

**Reply by yhmo** [reply]
@barzan-hayati 
To increase the ingestion speed, insert the rows with batch. In milvus, the default RPC transfer size limit for each insert request is 64MB. So, for dimension 512 vectors, 10000 rows per batch is a good choice.

One row for each insert() is not efficient. Some users might do this because their data comes from other sources by streaming.

Don't call flush() after each insert(). Normally, you don't need to call it manually, because milvus automatically manages flush actions. If you want to ensure all the data is persisted into S3, just call it once after all the data is inserted.

```
from pymilvus import (
    FieldSchema, CollectionSchema, DataType,
    utility, Collection, connections, Prepare
)
import random
import time

connections.connect("default", host="localhost", port=19530)

collection_name = "aaa"
dim = 512

schema = CollectionSchema(fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=dim),
])
utility.drop_collection(collection_name)
collection = Collection(collection_name, schema)
print("collection created")


batch = 10000
for i in range(50):
    data = [
        [i*batch+k for k in range(batch)], # id
        [[random.random() for _ in range(dim)] for _ in range(batch)], # vector
    ]
    res = collection.insert(data=data)
    print("batch", i)

collection.flush()
collection.create_index(field_name = "vector", index_params = {"index_type": "IVF_FLAT", "metric_type": "L2", "params": {"nlist": 64}})
collection.load()
print("collection loaded")

res = collection.query(expr="", output_fields=["count(*)"])
print(res) # get the row count
```

**Reply by barzan-hayati** [reply]
> One row for each insert() is not efficient. Some users might do this because their data comes from other sources by streaming.

Thanks @yhmo 

You're correct; in my situation, the new data arrives sporadically.
